### PR TITLE
Make KneeLRScheduler compatible PyTorch scheduler format

### PR DESCRIPTION
**Pull request by Animatory**
KneeLRScheduler now inherits _LRScheduler, supports momentum for Adam and SGD. Add non-zero start lr and non-zero final lr. Add different ways to set up a number of training steps as it is done in OneCycleLR.
Add documentation for the class.

### CODE REVIEW

**Comment by nikhil-iyer-97** [comment]
Mighty thanks! Let me spend a few hours going through this PR. Will approve it as soon as I test it on a few testcases and everything works out.

File: README.md [hunk 1]
```diff
@@ -14,7 +14,7 @@ This repo presents the wide minima density hypothesis as proposed in the followi
 ### Prerequisite: 
 * CUDA, cudnn
 * Python 3.6+
-* PyTorch 1.4.0
+* PyTorch 1.4+
```
Comment (nikhil-iyer-97) [inline README.md:17]: missed this, thanks!

File: README.md [hunk 2]
```diff
@@ -60,7 +60,8 @@ Based on the density of wide vs narrow minima , we propose the Knee LR schedule
 To use the Knee Schedule, import the scheduler into your training file:
 ```python
 >>> from knee_lr_schedule import KneeLRScheduler
->>> scheduler = KneeLRScheduler(optimizer, peak_lr, warmup_steps, explore_steps, total_steps)
+>>> scheduler = KneeLRScheduler(optimizer, max_lr, total_steps=total_steps,
```
Comment (nikhil-iyer-97) [inline README.md:63]: may have to change variable names, merging it with test branch for now

File: README.md [hunk 3]
```diff
@@ -73,13 +74,40 @@ To use it during training :
...
+- `max_lr` (float): The peak learning required for explore phase to escape narrow minimums.
+- `total_steps` (int): The total number of steps in the cycle. Note that if a value is not provided here, 
+  then it must be inferred by providing a value for epochs and steps_per_epoch.
+- `epochs` (int): The number of epochs to train for. This is used along with steps_per_epoch in order to infer 
+  the total number of steps in the cycle if a value for total_steps is not provided.
+- `steps_per_epoch` (int): The number of steps per epoch to train for. This is used along with epochs in order to 
+  infer the total number of steps in the cycle if a value for total_steps is not provided.
+- `pct_start` (float): The percentage of the total steps spent increasing the learning rate.
+- `pct_explore` (float): The percentage of the total steps spent on explore phase (keeping max the learning rate).
+- `cycle_momentum` (bool): If ``True``, momentum is cycled inversely
```
Comment (nikhil-iyer-97) [inline README.md:87]: not sure if we want to provide cyclical momentum . No idea how it will affect the explore phase and the effect it may have on reaching a wider minima

File: knee_lr_schedule.py [hunk 4]
```diff
@@ -1,36 +1,174 @@
...
-            param_group['lr'] = self.get_lr(self.current_step)
\ No newline at end of file
+import warnings
+from typing import Union, List
+
+from torch.optim.lr_scheduler import _LRScheduler
+from torch.optim.optimizer import Optimizer
+
+
+class KneeLRScheduler(_LRScheduler):
```
Comment (nikhil-iyer-97) [inline knee_lr_schedule.py:8]: importing _LRScheduler is a good idea, missed it out initially . Will have to change the args as we dont want users to pass in many parameters. Main parameters ( explore , total steps or total epochs , warmup steps). Remaining can be the same as other schedulers in pytorch.

File: knee_lr_schedule.py [hunk 5]
```diff
@@ -1,36 +1,174 @@
...
+    """
+    `Wide-minima Density Hypothesis and the Explore-Exploit Learning Rate Schedule`
```
Comment (nikhil-iyer-97) [inline knee_lr_schedule.py:10]: Thanks for citing the work

File: knee_lr_schedule.py [hunk 6]
```diff
@@ -1,36 +1,174 @@
...
+        last_epoch (int): The index of the last batch. This parameter is used when
+            resuming a training job. Since `step()` should be invoked after each
+            batch instead of after each epoch, this number represents the total
+            number of *batches* computed, not the total number of epochs computed.
+            When last_epoch=-1, the schedule is started from the beginning.
+        verbose (bool): If ``True``, prints a message to stdout for
+            each update.
+    """
+
+    def __init__(self, optimizer: Optimizer, max_lr: Union[List[float], float], total_steps: int = None,
```
Comment (nikhil-iyer-97) [inline knee_lr_schedule.py:57]: thanks for importing and using typing. Helps with documentation and readability

**Review by nikhil-iyer-97 — APPROVED** [review]
looks good, merging into test branch for now. Will take this up and make necessary edits

File: README.md (hunk 3 above)
Comment (Animatory) [reply README.md:87]: So, it will be better to set this parameter `False` by default

File: knee_lr_schedule.py (hunk 4 above)
Comment (Animatory) [reply knee_lr_schedule.py:8]: Actually, there are only 3-4 required args. The rest of args provide customizability of schedule

**Comment by Animatory** [comment]
Heard of you from the issue on PyTorch. Good job!

**Comment by nikhil-iyer-97** [comment]
Thank you!
